File: src/main/scala/edu/hitsz/aircraft/HeroAircraft.scala
package edu.hitsz.aircraft

import edu.hitsz.bullet.{AbstractBullet, HeroBullet}

import java.util

/**
 * 英雄飞机，游戏玩家操控
 *
 * @author hitsz
 */
class HeroAircraft(locationX: Int, locationY: Int, speedX: Int, speedY: Int, hpInit: Int)

/**
 * @param locationX 英雄机位置x坐标
 * @param locationY 英雄机位置y坐标
 * @param speedX    英雄机射出的子弹的基准速度（英雄机无特定速度）
 * @param speedY    英雄机射出的子弹的基准速度（英雄机无特定速度）
 * @param hpInit    初始生命值
 */ extends AbstractAircraft(locationX, locationY, speedX, speedY, hpInit) {
  /** 攻击方式 */
  private val shootNum = 1 //子弹一次发射数量
  private val power = 30 //子弹伤害
  private val direction = -1 //子弹射击方向 (向上发射：1，向下发射：-1)

  override def forward() = {
    // 英雄机由鼠标控制，不通过forward函数移动
  }

  override /*
      通过射击产生子弹
      @return 射击出的子弹List
     */ def shoot() = {
    val x = this.getLocationX
    val y = this.getLocationY + direction * 2
    val speedX = 0
    val speedY = this.getSpeedY + direction * 5
    var abstractBullet: HeroBullet = null
    for {i <- 0 until shootNum} yield new HeroBullet(x + (i * 2 - shootNum + 1) * 10, y, speedX, speedY, power)
  }
}

File: base/src/main/java/edu/hitsz/aircraft/HeroAircraft.java
package edu.hitsz.aircraft;

import edu.hitsz.animate.AnimateContainer;
import edu.hitsz.bullet.BaseBullet;
import edu.hitsz.bullet.HeroBulletFactory;
import edu.hitsz.prop.AbstractProp;
import edu.hitsz.vector.Vec2;

import java.util.LinkedList;

/**
 * 英雄飞机，游戏玩家操控
 *
 * @author hitsz
 */
public class HeroAircraft extends AbstractAircraft {
    /**
     * 子弹一次发射数量
     */
    private int shootNum = 10;

    public HeroAircraft(Vec2 posInit, AnimateContainer animateContainer, int hp) {
        super(posInit, animateContainer, hp, 0);
    }

    @Override
    public void forward() {
        // 英雄机由鼠标控制，不通过forward函数移动
    }

    /**
     * 通过射击产生子弹
     *
     * @return 射击出的子弹List
     */
    @Override
    public LinkedList<BaseBullet> shoot() {
        LinkedList<BaseBullet> res = new LinkedList<>();
        for (int i = 0; i < shootNum; i++) {
            // 子弹发射位置相对飞机位置向前偏移
            // 多个子弹横向分散
            BaseBullet baseBullet = new HeroBulletFactory(getPosition().plus(new Vec2((i * 2 - shootNum + 1) * 10, 0))).create();
            res.add(baseBullet);
        }
        return res;
    }

    @Override
    public LinkedList<AbstractProp> dropProps() {
        return new LinkedList<>();
    }

    public void increaseShootNum() {
        shootNum++;
    }
}
